File: internal/def/function.go
package def

type When uint

const (
	Activated   When = 0
	Deactivated When = 1
	Any         When = 2
)

type FunctionDefinition struct {
	Name        string
	Category    string
	Description string
	When        When
	Order       int
}

type FunctionKey struct {
	IsDouble  bool
	IsAlt     bool
	IsControl bool
	IsShift   bool
	IsWin     bool
	KeyCode   uint32
}

var nextOrder = 0

func nF(category, name, desc string, when ...When) (f *FunctionDefinition) {
	nextOrder++
	f = new(FunctionDefinition)
	f.Order = nextOrder
	f.Category = category
	f.Name = name
	f.Description = desc
	if len(when) == 0 {
		f.When = Activated
	} else {
		f.When = when[0]
	}
	FunctionDefinitions = append(FunctionDefinitions, f)
	FunctionNameMap[name] = f
	return
}

File: ts/src/util/function.ts
import { FunctionKey, getKeyText } from "../gobind";

export async function functionKeyToText(k: FunctionKey): Promise<string> {
  let s = "";

  if (k.IsWin) {
    s += "<W> + ";
  }

  if (k.IsControl) {
    s += "<C> + ";
  }

  if (k.IsAlt) {
    s += "<A> + ";
  }

  if (k.IsShift) {
    s += "<S> + ";
  }

  if (k.KeyCode !== 0) {
    s += await getKeyText(k.KeyCode);
  } else {
    s = s.substring(0, s.length - 3);
  }

  if (k.IsDouble) {
    s += " X2";
  }
  return s;
}
